res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+        }
+
+        // common functions
+        // In most cases, you should not overwrite these.
+        inGetParameter() {
+            return engine.getParameter(this.group, this.inKey);
+        }
+        inSetParameter(value) {
+            engine.setParameter(this.group, this.inKey, value);
+        }
+        inGetValue() {
+            return engine.getValue(this.group, this.inKey);
+        }
+        inSetValue(value) {
+            engine.setValue(this.group, this.inKey, value);
+        }
+        inToggle() {
+            this.inSetValue(!this.inGetValue());
+        }
+
+        outGetParameter() {
+            return engine.getParameter(this.group, this.outKey);
+        }
+        outSetParameter(value) {
+            engine.setParameter(this.group, this.outKey, value);
+        }
+        outGetValue() {
+            return engine.getValue(this.group, this.outKey);
+        }
+        outSetValue(value) {
+            engine.setValue(this.group, this.outKey, value);
+        }
+        outToggle() {
+            this.outSetValue(!this.outGetValue());
+        }
+
+        connect() {
+            /**
+            Override this method with a custom one to connect multiple Mixxx COs for a single Component.
+            Add the connection objects to the this.connections array so they all get disconnected just
+            by calling this.disconnect(). This can be helpful for multicolor LEDs that show a
+            different color depending on the state of different Mixxx COs. Refer to
+            SamplerButton.connect() and SamplerButton.output() for an example.
+            **/
+            if (this.group !== undefined &&
+                this.outKey !== undefined &&
+                this.output !== undefined &&
+                typeof this.output === "function") {
+                this.connections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+            }
+        }
+        disconnect() {
+            for (const conn of this.connections) {
+                conn.disconnect();
+            }
+        }
+        trigger() {
+
+            for (const conn of this.connections) {
+                conn.trigger();
+            }
+        }
+        send(value) {
+            if (this.midi === undefined || this.midi[0] === undefined || this.midi[1] === undefined) {
+                return;
+            }
+            midi.sendShortMsg(this.midi[0], this.midi[1], value);
+            if (this.sendShifted) {
+                if (this.shiftChannel) {
+                    midi.sendShortMsg(this.midi[0] + this.shiftOffset, this.midi[1], value);
+                } else if (this.shiftControl) {
+                    midi.sendShortMsg(this.midi[0], this.midi[1] + this.shiftOffset, value);
+                }
+            }
+        }
+    };
+    Component.prototype.max = 127;
+    Component.prototype.shiftOffset = 0;
+    Component.prototype.sendShifted = false;
```

Comment:
I believe this means:  // if true, output values are sent not only to the designated midi channel and control, but also sent to associated midi controls. See below
// if shiftChannel is set, output values will be sent to the regular midi channel as well as the channel offset by this amount
// if shiftControl is set, output values will be sent to the regular midi control as well as the control offset by this amount.

Reply:
You got it right. Sure I can leave comments, though are they really needed? `send()`  is so simple the code is pretty much self documenting imo.

Reply:
At least for myself, when I read a library API I'd prefer not to have to read the code to know how to use it.  My personal reference for best documentation is the google Abseil library ([example](https://github.com/abseil/abseil-cpp/blob/master/absl/time/civil_time.h)).  The comments are extensive and give lots of usage examples.  I'm not saying we have to do that much, but a few plain words in a comment can make it much easier for a programmer to understand what to do.  Remember that this particular library is meant to be used by many developers who are not familiar with Mixxx at all, so I believe we should do a little more work here to make it accessible to them.

Reply:
Well, we already have extensive documentation in the form of the wiki page. Having the wiki page generated from docs and examples within the code would be great, but we don't have the facilities for that right now. Maybe jsdoc can help. I'll look into whether this is possible.

Reply:
Generating online documentation from inline comments would be nice but beyond the scope of this PR.